File: on-premise-asr-code-samples/javascript/streaming/client-file/streaming-client-file.mjs
#!/usr/bin/env node

/*
 * A websocket client for banafo-streaming-server
 *
 * Usage:
 *     ./streaming-client-file.mjs \
 *       --addr localhost \
 *       --port 6006 \
 *       --seconds-per-message 0.1 \
 *       --samples-per-message 8000 \
 *       /path/to/foo.wav
 *
 * (Note: You have to first start the server before starting the client)
 */

import fsPromises from "node:fs/promises";
import { WebSocket } from "ws";
import wavefile from "wavefile";
import { hideBin } from "yargs/helpers";
import yargs from "yargs/yargs";

const getArgs = () => {
  return yargs(hideBin(process.argv))
    .scriptName("streaming-client-file.mjs")
    .command(
      "$0 [addr] [port] [samples-per-message] [seconds-per-message] <file>",
      "",
    )
    .option("addr", {
      type: "string",
      default: "localhost",
      description: "Address of the server",
    })
    .option("port", {
      type: "number",
      default: 6006,
      description: "Port of the server",
    })
    .option("samples-per-message", {
      type: "number",
      default: 8000,
      description: "Number of samples per message",
    })
    .option("seconds-per-message", {
      type: "number",
      default: 0.1,
      description:
        "We will simulate that the duration of two messages is of this value",
    })
    .positional("file", {
      type: "string",
      description:
        "Input WAV file to decode (single channel, 16 kHz sample rate)",
    })
    .strict()
    .wrap(yargs().terminalWidth())
    .parse();
};

const sleep = (seconds) => {
  return new Promise((resolve) => {
    setTimeout(resolve, seconds * 1000);
  });
};

const readWav = async (file) => {
  const f = new wavefile.WaveFile(await fsPromises.readFile(file));

  if (f.fmt.sampleRate !== 16000) {
    throw new Error("Only 16 kHz sample rate supported");
  }

  if (f.fmt.numChannels !== 1) {
    throw new Error("Only mono audio files are supported");
  }

  f.toBitDepth("32f");
  return f.getSamples(false, Float32Array);
};

const transcribe = async (
  addr,
  port,
  file,
  samplesPerMessage,
  secondsPerMessage,
) => {
  let websocket;

  try {
    const samples = await readWav(file);

    websocket = new WebSocket(`ws://${addr}:${port}`);

    console.info(`Sending ${file}`);

    const donePromise = new Promise((resolve) => {
      websocket.onmessage = (ev) => {
        const message = ev.data;
        if (message === "Done!") {
          resolve();
        } else {
          console.log(message);
        }
      };
    });

    await new Promise((resolve, reject) => {
      websocket.once("open", () => resolve());
      websocket.once("close", () => reject(new Error("Connection failure")));
    });

    let start = 0;
    while (start < samples.length) {
      let end = start + samplesPerMessage;
      end = Math.min(end, samples.length);
      const d = samples.subarray(start, end);

      websocket.send(d);

      // Simulate streaming. You can remove the sleep if you want
      await sleep(secondsPerMessage); // in seconds

      start += samplesPerMessage;
    }

    // to signal that the client has sent all the data
    websocket.send("Done");

    await donePromise;
  } finally {
    websocket?.close();
  }
};

const main = async () => {
  const args = getArgs();

  const addr = args.addr;
  const port = args.port;
  const file = args.file;
  const samplesPerMessage = args.samplesPerMessage;
  const secondsPerMessage = args.secondsPerMessage;

  await transcribe(addr, port, file, samplesPerMessage, secondsPerMessage);
};

main();

File: on-premise-asr-code-samples/python/streaming/client-file/streaming-client-file.py
#!/usr/bin/env python3

"""
A websocket client for banafo-streaming-server

Usage:
    ./streaming-client-file.py \
      --addr localhost \
      --port 6006 \
      --seconds-per-message 0.1 \
      --samples-per-message 8000 \
      /path/to/foo.wav

(Note: You have to first start the server before starting the client)
"""

import argparse
import asyncio
import logging
import wave
import websockets
import numpy as np


def read_wav(file):
    with wave.open(file) as f:
        assert f.getframerate() == 16000, f.getframerate()
        assert f.getnchannels() == 1, f.getnchannels()
        assert f.getsampwidth() == 2, f.getsampwidth()
        frames = f.readframes(f.getnframes())
        samples = np.frombuffer(frames, dtype=np.int16).astype(np.float32) / 32768
        return samples


def get_args():
    parser = argparse.ArgumentParser(
        formatter_class=argparse.ArgumentDefaultsHelpFormatter
    )

    parser.add_argument(
        "--addr",
        type=str,
        default="localhost",
        help="Address of the server",
    )

    parser.add_argument(
        "--port",
        type=int,
        default=6006,
        help="Port of the server",
    )

    parser.add_argument(
        "--samples-per-message",
        type=int,
        default=8000,
        help="Number of samples per message",
    )

    parser.add_argument(
        "--seconds-per-message",
        type=float,
        default=0.1,
        help="We will simulate that the duration of two messages is of this value",
    )

    parser.add_argument(
        "file",
        type=str,
        help="Input WAV file to decode (single channel, 16-bit int, 16 kHz sample rate)",
    )

    return parser.parse_args()


async def receive_results(socket):
    last_message = ""
    async for message in socket:
        if message != "Done!":
            last_message = message
            logging.info(message)
        else:
            break
    return last_message


async def transcribe(
    addr,
    port,
    file,
    samples_per_message,
    seconds_per_message,
):
    samples = read_wav(file)

    async with websockets.connect(
        f"ws://{addr}:{port}"
    ) as websocket:
        logging.info(f"Sending {file}")

        receive_task = asyncio.create_task(receive_results(websocket))

        start = 0
        while start < samples.shape[0]:
            end = start + samples_per_message
            end = min(end, samples.shape[0])
            d = samples.data[start:end].tobytes()

            await websocket.send(d)

            # Simulate streaming. You can remove the sleep if you want
            await asyncio.sleep(seconds_per_message)  # in seconds

            start += samples_per_message

        # to signal that the client has sent all the data
        await websocket.send("Done")

        decoding_results = await receive_task


async def main():
    args = get_args()
    logging.info(vars(args))

    addr = args.addr
    port = args.port
    file = args.file
    samples_per_message = args.samples_per_message
    seconds_per_message = args.seconds_per_message

    await transcribe(
        addr=addr,
        port=port,
        file=file,
        samples_per_message=samples_per_message,
        seconds_per_message=seconds_per_message,
    )


if __name__ == "__main__":
    fmt = "%(asctime)s %(levelname)s [%(filename)s:%(lineno)d] %(message)s"
    logging.basicConfig(format=fmt, level=logging.INFO)
    asyncio.run(main())
